// xsupport/main/Xlinode.js
const Joi = require("joi");
const Xfetch = require("./Xfetch");
const Xtime = require("./Xtime");

class Xlinode {
    constructor(user, pass) {
        let { error } = Joi.object({
            user: Joi.string().required(),
            pass: Joi.string().required()
        }).validate({ user, pass });
        if (error) throw error;

        this.show = false;

        this.fetch = new Xfetch(user);
        this.fetch.data = { user, pass };

        this.fetch.filter.values = {
            csrf_token: 'csrf_token'
        }
    }

    login() {
        this.#show(this.login);
        return this.fetch.get('https://login.linode.com/csrf')
            .then(_ => {
                let body = Xfetch.parseBody({
                    return_to:'https://login.linode.com/',
                    created:Xtime.format('YYYY-MM-DDTHH:mm:ss.SSSSSS'),
                    username:this.fetch.data.user,
                    password:this.fetch.data.pass,
                    remember:'y',
                    csrf_token:this.fetch.data.csrf_token,
                    submit:'Log in'
                })
                return this.fetch.post('https://login.linode.com/login',body);
            })
            .then(_=>!this.fetch.response.includes('Username or password incorrect'))
    }

    #show(text) {
        if (this.show) console.log(text);
    }
}


module.exports = Xlinode;
