// xsupport/main/Xproxyxoay.js
const Joi = require("joi");
const Xfetch = require("./Xfetch");
const Xproxy = require("./Xproxy");
const Xtime = require("./Xtime");


class Xproxyxoay {

    constructor(apikey) {
        let { error } = Joi.string().required().validate(apikey);
        if (error) throw error;

        this.apikey = apikey;
        this.baseUrl = 'https://proxyxoay.shop/api/get.php';
        this.fetch = new Xfetch();
        this.proxy = {
            px: null,
            text: '',
            agent: null,
            worker: 0,
            renew: true
        };
        this.autoChange = false;
        this.show = false;
    }

    getNewProxy() {
        return this.fetch.get(`${this.baseUrl}?key=${this.apikey}&tinhthanh=3`)
            .then(rs => {
                if (rs.status == 100) {
                    this.proxy.renew = false;
                    this.proxy.text = rs.proxyhttp;
                    this.proxy.px = new Xproxy('http', ...rs.proxyhttp.split(':'));
                    this.proxy.agent = this.proxy.px.agent();
                }
                if (this.show) console.log(`Get new proxy: ${this.proxy.text}\tRenew:${this.proxy.renew}`);
                return this.proxy;
            }).catch((err) => this.proxy)
    }

    async #autoChangeProxy(callback) {
        while (this.autoChange) {
            await this.#waitForWorkers();
            await this.getNewProxy();
            try { if (callback) callback(); } catch { }
            await Xtime.sleep(70000);
        }
    }

    async #waitForWorkers() {
        this.proxy.renew = true;
        while (this.proxy.worker > 0) {
            // if (this.show) console.log(`Đang đợi worker hoàn thành !`);
            await Xtime.sleep(1000);
        }
    }
    async waitRenew() {
        while (this.proxy.renew) {
            await Xtime.sleep(1000);
        }
    }

    autoChangeProxy(callback) {
        this.autoChange = true;
        return this.#autoChangeProxy(callback);
    }
    stopAutoChangeProxy() {
        this.autoChange = false;
    }
}

module.exports = Xproxyxoay;
